demonstrations/tutorial_optimal_control.py

```diff
@@ -0,0 +1,847 @@
+        if isinstance(label, str):
+            label = f"${label[-1]}_{op.wires[0]}$"
+            dash = dashes[label[1]]
+        else:
+            label = "$" + " ".join([f"{n[-1]}_{w}" for w, n in zip(op.wires, label)]) + "$"
+            dash = [10, 0]
+
+        # Set color according to qubit the term acts on
+        col = colors[op.wires[0]]
+        # Plot the pulse
+        values = [pulse_fn(p, t) for t in plot_times]
+        ax.plot(plot_times, values, label=label, dashes=dash, color=col)
+    ax.legend()
+    # Set legends and axis descriptions
+    axs[0].legend(title="Single-qubit terms", ncol=int(jnp.sqrt(len(ops))))
+    axs[1].legend(title="Two-qubit terms")
+    title = f"{target_name}, Fidelity={max_profit:.6f}"
+    axs[0].set(ylabel=r"Pulse function $f(p, t)$", title=title)
+    axs[1].set(xlabel="Time $t$", ylabel=r"Pulse function $f(p, t)$")
+    plt.show()
+
+
+plot_optimal_pulses(hist, f, ops_param, T, target_name)
+
+#############################################################################
+# We observe that a single rectangular pulse is sufficient for most of the
+# generating terms in the Hamiltonian, and we see that their shape is closer to
+# actual rectangles now, in particular for those with a saturated amplitude.
+#
+# The final fidelity tells us that we achieved our goal of finding a pulse
+# sequence that implements a unitary close to a CNOT gate.
+# It could be optimized further, for example by running the optimization for more
+# training iterations, by tuning the optimizer further to avoid oscillations,
+# or by increasing the precision with which we run the ODE solver.
+#
+# Pulse sequence for Toffoli
+# --------------------------
+#
+# The second example we consider is the compilation of a Toffoli--or CCNOT--gate.
+# We reuse most of the workflow from above and only change the pulse Hamiltonian as
+# well as a few hyperparameters.
+#
+# In particular, the Hamiltonian uses the drift term :math:`H_d=Z_0+Z_1+Z_2`
+# and the generators are all single-qubit Pauli operators on all three qubits, together
+# with the interaction generators :math:`Z_0X_1, Z_1X_2, Z_2X_0`. Again,
+# all parametrized terms use the coefficient function ``smooth_rectangles``.
+# We allow for a longer pulse duration of :math:`3\pi` and five smooth rectangles in
+# each pulse shape.
+#
+# In summary, we use nine single-qubit generators and three two-qubit generators, with
+# five rectangles in each pulse shape and each rectangle being given by an amplitude and
+# a start and end time. The pulse sequence thus has :math:`(9+3)\cdot 5\cdot 3=180`
+# parameters.
+
+num_wires = 3
+# New pulse hyperparameters
+T = 3 * jnp.pi  # Longer total duration
+eps = 0.1 * T
+P = 5  # More rectangles in sum: P=5
+f = partial(smooth_rectangles, k=k, max_amp=max_amp, eps=eps, T=T)
+
+# Hamiltonian terms of the drift and parametrized parts of H
+ops_H_d = [Z(0), Z(1), Z(2)]
+ops_param = [P(w) for P in [X, Y, Z] for w in range(num_wires)]
```

Comment:
Maybe we should avoid using `P` here since it is the number of rectangles in every sum.
```suggestion
ops_param = [pauli(w) for pauli in [X, Y, Z] for w in range(num_wires)]
```
I am not very happy with the `pauli` name. It's mainly to avoid any possible confusion. Feel free to use any other name 😄 Again, not too critical.

Reply:
Went for `pauli_op` :)

Reply:
Much better!